### ISSUE TITLE
RFMIP SW standalone test works on GPUs on Summit but not other hosts

### ISSUE THREAD

**Issue by RobertPincus** [RobertPincus/rte-rrtmgp#40]
At present, using GPU code and running on GPUs produces correct answers for the RFMIP SW standalone test case on the OLCF host Summit but incorrect answers on other hosts. We are seeking a solution.

**Comment by RobertPincus** [comment]
@mrnorman had some suggestions: 

There are a few paths you can take. The easiest is to just turn on Managed memory. It should override all the data statements, and if managed memory works but it doesn't work without, you have a data problem either in the code or in the OpenACC runtime.

Something to keep in mind is that Summit supports accessing CPU data from the GPU but almost no other machine does. So if there's a sin you're committing there, it'll show up on another machine but not Summit. I have a hunch this is the problem, but whether it's you or it's the OpenACC runtime bugging out, I don't know. You'll recall I've had *lots* of problems with pointers doing the wrong thing in the OpenACC runtime. Your best bet might be telling them to simply turn on managed memory 🙂. It usually solves the problems outright.

For debugging, you have several options, ordered from easier to harder:

(1) Turn on "PCAST", which gives automatically dumps out diffs between redundant CPU and GPU code every time you "copyout". Depending on how much you copy out, this can be cumbersome to sift thorugh, but it's super easy to do.

https://www.pgroup.com/blogs/posts/pcast.htm

(2) Run cuda-memcheck, and see if it chokes on something. Narrow down where that is, and you're good to go.

(3) You can try running valgrind on the code, but...you do "interesting" things in Fortran, so valgrind will likely choke on those things (like auto-resizing array). It's worth a try, though.

(4) Change all data statements to default(present). PGI has a flag that's supposed to do this for you "-acc=defpresent". Try it out and see. What this will do is trigger an error if anything is not present or partially present. This can help with identifying times when data isn't where you thought it would be (sometimes due to OpenACC runtime not handling pointer mapping correctly).

**Comment by mrnorman** [comment]
> You can try running valgrind on the code, but...you do "interesting" things in Fortran
That was supposed to be just for you, Robert :)

**Comment by RobertPincus** [comment]
> > You can try running valgrind on the code, but...you do "interesting" things in Fortran
> 
> That was supposed to be just for you, Robert :)

Yeah, but I've got no shame and plenty of dirty laundry...

**Comment by mrnorman** [comment]
Don't we all

**Comment by RobertPincus** [comment]
@mrnorman I recompiled with -ta=tesla:cc60,cuda9.2,managed on Piz Daint and still get incorrect answers. Does that change your interpretation that it's a data statement issue?

**Comment by phmarti** [comment]
@RobertPincus I added a few missing data movements to get the code running with `-acc=defpresent`. Unfortunately, it still gives incorrect answers. Made a small PR #41 for the changes but It's quick and dirty as it doesn't solve the issue yet.

**Comment by mrnorman** [comment]
That is disappointing. I don't know what it changes because these things can be shifty. It could be a good old-fashioned compiler bug in the PTX output for cc60? It's hard to tell. Let me try your current code on my desktop at work, which has K40s (cc35) just to have another data point. It might take a day or two to get to this.

**Comment by mrnorman** [comment]
In the meantime, I suggest running PCAST to try to find exactly which kernel is diverging from the CPU.

**Comment by RobertPincus** [comment]
All, I've started a new branch to work on this issue. Please target gpu-platform-dependence with all PRs. 

@mrnorman I did use PCAST briefly; it seemed liked the adding kernel was potentially a problem. I'll look in more detail but likely not for a few days.

**Comment by naromero77** [comment]
@RobertPincus @mrnorman Is the issue with the develop branch? The CPU version works of course, but the GPU version gives me:

 ```
 PANDAS_TYPES = (pd.Series, pd.DataFrame, pd.Panel)
Downloading reference results
Variable rlu differs (max abs difference: 5.084980e+02; max frac. difference: 2.000000e+02%)
Variable rld differs (max abs difference: 4.628228e+02; max frac. difference: 2.000000e+02%)
Variable rsu differs (max abs difference: 3.202381e+02; max frac. difference: 2.000000e+02%)
Variable rsd differs (max abs difference: 1.326000e+03; max frac. difference: 2.000000e+02%)

```

Is this what you are seeing?

**Comment by RobertPincus** [comment]
@naromero77 This is extra-bad wookie. On Piz Daint and Mistral we have correct answers for the LW (relative differences of order 10e-5) and order 1 errors in the SW. Your results are incorrect for both codes. 

Where are you working, with what compiler and version?

**Comment by mrnorman** [comment]
If you're using the default staged baseline files, they were stale last time I tried, and Robert was going to update them.

**Comment by RobertPincus** [comment]
@mrnorman The last commit to master, ed5b011 from a week ago, updates the code to be consistent with the baseline files. 

I'd like to work from branch gpu-platform-dependence on this issue. It differs from master so far only in the new data statements added by @phmarti.

**Comment by mrnorman** [comment]
Cool, sorry for the stale comment.

**Comment by naromero77** [comment]
> @naromero77 This is extra-bad wookie. On Piz Daint and Mistral we have correct answers for the LW (relative differences of order 10e-5) and order 1 errors in the SW. Your results are incorrect for both codes.
> 
> Where are you working, with what compiler and version?

@RobertPincus I am working on my Linux desktop that is administered by Argonne National Laboratory.  It uses PGI 19.3 and here is the output from `nvidia-smi`
```
(e3sm) [naromero@naromero-desktop ~]$ nvidia-smi
Wed Aug 14 12:00:51 2019       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.39       Driver Version: 418.39       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 105...  Off  | 00000000:0F:00.0  On |                  N/A |
| 30%   26C    P8    N/A /  75W |    480MiB /  4031MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1580      G   /usr/bin/X                                   217MiB |
|    0      2196      G   ...quest-channel-token=8648259451609102894    85MiB |
|    0      3072      G   compiz                                       174MiB |
+-----------------------------------------------------------------------------+
```
I am using these flags:
export FCFLAGS=`-O3 -g -ta=tesla -Minfo=accel`

**Comment by naromero77** [comment]
@RobertPincus I will try the `gpu-platform-dependence` branch next.

**Comment by naromero77** [comment]
@RobertPincus No change, still those huge differences.

**Comment by RobertPincus** [comment]
@naromero77 Our testing on Summit has been done with PGI 19.4. Do you have this available? 

On Summit @mrnorman has been using -O0, on Daint we have been using no optimization at all. This might also be worth checking.

**Comment by naromero77** [comment]
@RobertPincus I re-ran with PGI 19.4 and compiled with -O0. No change whatsoever. I must be doing something different to get a 200% difference.

When I ran on Summit a while back it worked with GPUs it did work, though I had to re-generate the CPU reference again on Summit with the PGI compiler.

I need to go out of town for a while and will be back on Monday. Are there any obvious things to try?

**Comment by RobertPincus** [comment]
@naromero77 Since I don't have access to a system like yours I'm not sure how to help, especially since this works for CPUs but not for GPUs. Can you work with a local GPU expert like @mrnorman ?

**Comment by naromero77** [comment]
@RobertPincus I am back in town and will work on this. I had the admins install PGI 19.7 and re-compiling libraries with Spack. I have Pascal GPU (cc60) and will try PCAST to narrow things down before I bug Matt.

**Comment by naromero77** [comment]
@RobertPincus Looks like this was a stupid mistake on my part, but there are still some differences that I would like you to review. Problem was that `RTE_KERNEL=openacc` was not being properly exported in my script.

This is what I am getting now:
```
Downloading reference results
Variable rlu: No diffs
Variable rld differs (max abs difference: 1.490116e-08; max frac. difference: 1.173428e-05%)
Variable rsu differs (max abs difference: 3.051758e-05; max frac. difference: 1.184619e-05%)
Variable rsd differs (max abs difference: 6.103516e-05; max frac. difference: 1.087066e-05%)
```

I also ran PCAST with a relative tolerance of 1e-05. There is something that differs by inf. I am attaching the output of pcast for the two runs.

[sw_pcast.out.txt](https://github.com/RobertPincus/rte-rrtmgp/files/3527600/sw_pcast.out.txt)
[lw_pcast.out.txt](https://github.com/RobertPincus/rte-rrtmgp/files/3527601/lw_pcast.out.txt)

**Comment by naromero77** [comment]
@mrnorman Is there a list of OpenACC directives that can cause PCAST to give a false positive?

**Comment by RobertPincus** [comment]
@naromero77 The differences you are seeing are consistent with changes across compilers, platforms, etc. That suggests the code is in fact working correctly, modulo what we learn from PCAST.

**Comment by naromero77** [comment]
@RobertPincus My first thought is that what I am seeing with `pcast` is a false positive. Based on my interpretation of the PGI documentation, the canonical example that PCAST is supposed to work on is when you copy things in and out of main memory. However, there are a number of intermediate data structures that only get created/initialized on the GPU (using `acc created`, e.g. this tropo array). Tomorrow I will create a self-standing case to test my hypothesis.

**Comment by mrnorman** [comment]
@naromero77 , One thing that can cause problems is !$acc create(). If you copyout() or update host() something that was created but not yet written to, for instance, it can lead to different values between the redundant execution on CPU and GPU.

**Comment by naromero77** [comment]
@mrnorman @RobertPincus This is what I am trying to do, but have not had much luck yet. There is array in `mo_gas_optics_kernels.F90` gets identified as potentially problematic,
```
logical(wl), dimension(ncol,nlay), intent(out) :: tropo
```

I tried initializing to true on the CPU and then doing a copyin. I always get this output from PCAST:
```
	idx: 45000 FAIL act: 1 exp: 4294967295
	idx: 45001 FAIL act: 1 exp: 4294967295
	idx: 45002 FAIL act: 1 exp: 4294967295
	idx: 45003 FAIL act: 1 exp: 4294967295
	idx: 45004 FAIL act: 1 exp: 4294967295
	idx: 45005 FAIL act: 1 exp: 4294967295
	idx: 45006 FAIL act: 1 exp: 4294967295
	idx: 45007 FAIL act: 1 exp: 4294967295
	idx: 45008 FAIL act: 1 exp: 4294967295
	idx: 45009 FAIL act: 1 exp: 4294967295
	idx: 45010 FAIL act: 1 exp: 4294967295
	idx: 45011 FAIL act: 1 exp: 4294967295
	idx: 45012 FAIL act: 1 exp: 4294967295
	idx: 45013 FAIL act: 1 exp: 4294967295
	idx: 45014 FAIL act: 1 exp: 4294967295
	idx: 45015 FAIL act: 1 exp: 4294967295
	idx: 45016 FAIL act: 1 exp: 4294967295
	idx: 45017 FAIL act: 1 exp: 4294967295
	idx: 45018 FAIL act: 1 exp: 4294967295
	idx: 45019 FAIL act: 1 exp: 4294967295
	idx: 45020 FAIL act: 1 exp: 4294967295
	idx: 45021 FAIL act: 1 exp: 4294967295
	idx: 45022 FAIL act: 1 exp: 4294967295
	idx: 45023 FAIL act: 1 exp: 4294967295
	idx: 45024 FAIL act: 1 exp: 4294967295
	idx: 45025 FAIL act: 1 exp: 4294967295
```
I tried creating a simple isolated test case to see if PCAST had a problem with the `logical` data type, but it seems to work (as far as the compiler is concerned, it seems to treat it as an `unsigned int`.

Tropo is computed as follows:
```
! determine if in lower or upper part of atmosphere
        tropo(icol,ilay) = log(play(icol,ilay)) > press_ref_trop_log
```
play is copyin from the CPU as far as I can tell.

**Comment by RobertPincus** [comment]
For anyone following along at home: this issue is related to compilation, not code, on at least some platforms (e.g. Piz Daint from CSCS). The default wrappers mix llvm and non-llvm libraries during linking. On Piz Daint specifically the code produces correct answers using `pgf90` instead of `ftn`.

**Comment by RobertPincus** [comment]
Fixed by pull request #44.
